### PR TITLE
feat!: add Content Source Map functionality

### PR DESCRIPTION
**Pull request by stipsan**

### CODE REVIEW

**Comment by kmelve** [comment]
#### 📝 Changed routes:


| [**/**](https://nextjs-blog-cms-sanity-v3-glg4xs52e.sanity.build/) | [**/posts/\[slug\]**](https://nextjs-blog-cms-sanity-v3-glg4xs52e.sanity.build/posts/[slug]) |
|:-:|:-:|
|<a href="https://nextjs-blog-cms-sanity-v3-glg4xs52e.sanity.build/">Screenshot of /</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/rBmjaazq8hGiyQ9Iydc3mXiefuRRvkGOvl2iiL5gDJ9oEvVvfsXQ3V5mz9SWznH3oQvpJLVwN509vPfC8zohLg.png">(view full size)</a> |<a href="https://nextjs-blog-cms-sanity-v3-glg4xs52e.sanity.build/posts/[slug]">Screenshot of /posts/[slug]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/FHBR99_Q-BU3WHDnvOA5Wr3_GXpiHFse48yS2q7C8W1wzd1dAwURRapfb6ih1OkaFnnvlt5ZhQiCX-6_LwlI_Q.png">(view full size)</a> |



And 3 other routes:
- [**/api/og**](https://nextjs-blog-cms-sanity-v3-glg4xs52e.sanity.build/api/og)
- [**/api/preview**](https://nextjs-blog-cms-sanity-v3-glg4xs52e.sanity.build/api/preview)
- [**/api/revalidate**](https://nextjs-blog-cms-sanity-v3-glg4xs52e.sanity.build/api/revalidate)



Commit a95a9b5a02735f88e64976aa6fd82f25933fedfe (https://nextjs-blog-cms-sanity-v3-glg4xs52e.sanity.build).

File: lib/sanity.api.ts [hunk 1]
```diff
@@ -19,3 +19,6 @@ export const useCdn = process.env.SANITY_REVALIDATE_SECRET
 // This is the document id used for the preview secret that's stored in your dataset.
 // The secret protects against unauthorized access to your draft content and have a lifetime of 60 minutes, to protect against bruteforcing.
 export const previewSecretId: `${string}.${string}` = 'preview.secret'
+
+// THe route path to where the Studio is embedded
```
Comment (snorrees) [inline lib/sanity.api.ts:23]:
```suggestion
// The route path to where the Studio is embedded
```

File: lib/sanity.client.ts [hunk 2]
```diff
@@ -13,51 +19,68 @@ import { createClient } from 'next-sanity'
...
-const client = projectId
-  ? createClient({ projectId, dataset, apiVersion, useCdn })
+const sanityClient = projectId
+  ? createClient({
+      projectId,
+      dataset,
+      apiVersion,
+      useCdn,
+      studioUrl: studioBasePath,
+      // @TODO stop encoding source maps in development after finishing the migration
```
Comment (snorrees) [inline lib/sanity.client.ts:29]: Stale todo?

**Review by snorrees — APPROVED** [review]

File: components/PreviewProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import { dataset, projectId, studioBasePath } from 'lib/sanity.api'
+import { createClient } from 'next-sanity'
+import { LiveStoreProvider } from 'next-sanity/preview/live-store'
+import { useMemo } from 'react'
+
+/**
+ * LiveStoreProvider requires Content Source Maps in order to preview in real-time, as you type, changes as they happen.
+ * If Visual Editing isn't used, and the amount of documents isn't in the tens of thousands then:
+ * import {GroqStoreProvider} from 'next-sanity/preview-groq-store'
```
[the remark this answers is not in the document]
Comment (stipsan) [reply components/PreviewProvider.tsx:9]: Ooops didn't realise this PR wasn't market as draft. I'm still working on this and trying to find a good way to switch between the modes. And find a good balance between a live preview that works for everyone, and a live preview that still works once visual editing is enabled 🙌 
Setting the PR back to draft

**Pull request closed without merging**